src-ghc/Pact/Server/ApiServer.hs

```diff
@@ -156,14 +155,21 @@ buildCmdRpc c@Command {..} = do
   log $ "Processing command with hash: " ++ show _cmdHash
   return (cmdToRequestKey c,fmap encodeUtf8 c)
 
-group :: Int -> [a] -> [[a]]
-group _ [] = []
+-- | Reorganize a `NonEmpty` into sub-batches.
+group :: Int -> NonEmpty a -> NonEmpty (NonEmpty a)
 group n l
-  | n > 0 = take n l : group n (drop n l)
-  | otherwise = error "Negative n"
+  | n < 1 = error "Non-positive n"
+  | otherwise = case rest of
+      Nothing -> items :| []
+      Just rs -> items <| group n rs
+  where
+    (items, rest) = bimap NEL.fromList NEL.nonEmpty $ NEL.splitAt n l
 
-queueCmds :: (MonadReader ApiEnv m, MonadIO m) => [(RequestKey,Command ByteString)] -> m [RequestKey]
+queueCmds :: (MonadReader ApiEnv m, MonadIO m) =>
+  NonEmpty (RequestKey, Command ByteString) -> m (NonEmpty RequestKey)
 queueCmds rpcs = do
   hc <- view aiHistoryChan
-  liftIO $ writeHistory hc $ AddNew (map snd rpcs)
-  return $ fst <$> rpcs
+  liftIO . writeHistory hc . AddNew $ NEL.toList cmds
```

Comment:
Should `AddNew` take a `NonEmpty` so we can force the list conversion down into the db layer?

Reply:
I'd personally say yes, but didn't want to spread `NonEmpty` further than I had to.